src/valibot/index.ts

```diff
@@ -0,0 +1,296 @@
+    const arrayGen = `v.array(${maybeLazy(type.type, gen)})`;
+    if (!isNonNullType(parentType)) {
+      return `v.nullish(${arrayGen})`;
+    }
+    return arrayGen;
+  }
+  if (isNonNullType(type)) {
+    const gen = generateFieldTypeValibotSchema(config, visitor, field, type.type, type);
+    return maybeLazy(type.type, gen);
+  }
+  if (isNamedType(type)) {
+    const gen = generateNameNodeValibotSchema(config, visitor, type.name);
+    if (isListType(parentType))
+      return `v.nullable(${gen})`;
+
+    const actions = actionsFromDirectives(config, field);
+
+    if (isNonNullType(parentType))
+      return pipeSchemaAndActions(gen, actions); ;
+
+    return `v.nullish(${pipeSchemaAndActions(gen, actions)})`;
+  }
+  console.warn('unhandled type:', type);
+  return '';
+}
+
+function actionsFromDirectives(config: ValidationSchemaPluginConfig, field: InputValueDefinitionNode | FieldDefinitionNode): string[] {
+  if (config.directives && field.directives) {
+    const formatted = formatDirectiveConfig(config.directives);
+    return buildApiForValibot(formatted, field.directives);
+  }
+
+  return [];
+}
+
+function pipeSchemaAndActions(schema: string, actions: string[]): string {
+  if (actions.length === 0)
+    return schema;
+
+  return `v.pipe(${schema}, ${actions.join(', ')})`;
+}
+
+function generateNameNodeValibotSchema(config: ValidationSchemaPluginConfig, visitor: Visitor, node: NameNode): string {
+  const converter = visitor.getNameNodeConverter(node);
+
+  switch (converter?.targetKind) {
+    case 'InterfaceTypeDefinition':
+    case 'InputObjectTypeDefinition':
+    case 'ObjectTypeDefinition':
+    case 'UnionTypeDefinition':
+      // using switch-case rather than if-else to allow for future expansion
+      switch (config.validationSchemaExportType) {
+        default:
+          return `${converter.convertName()}Schema()`;
+      }
+    case 'EnumTypeDefinition':
+      return `${converter.convertName()}Schema`;
+    case 'ScalarTypeDefinition':
+      return valibot4Scalar(config, visitor, node.value);
+    default:
+      if (converter?.targetKind)
+        console.warn('Unknown targetKind', converter?.targetKind);
+
+      return valibot4Scalar(config, visitor, node.value);
+  }
+}
+
+function maybeLazy(type: TypeNode, schema: string): string {
+  if (isNamedType(type) && isInput(type.name.value))
+    return `v.lazy(() => ${schema})`;
+
+  return schema;
+}
+
+function valibot4Scalar(config: ValidationSchemaPluginConfig, visitor: Visitor, scalarName: string): string {
+  if (config.scalarSchemas?.[scalarName])
+    return config.scalarSchemas[scalarName];
+
+  const tsType = visitor.getScalarType(scalarName);
+  switch (tsType) {
+    case 'string':
+      return `v.string()`;
+    case 'number':
+      return `v.number()`;
+    case 'boolean':
+      return `v.boolean()`;
+  }
+  console.warn('unhandled scalar name:', scalarName);
+  return 'v.any()';
```

Comment:
Since no type errors occurred, `definedNonNullAny` was implemented without using it.